app/src/main/kotlin/compiler/syntax/AstFactory.kt

```diff
@@ -50,6 +52,25 @@ object AstFactory {
         return (properNode.children.first() as ParseTree.Leaf).content
     }
 
+    private fun extractForeignName(parseTree: ParseTree.Leaf<Symbol>, diagnostics: Diagnostics, asIdentifier: Boolean): String {
+        return when (parseTree.token()) {
+            TokenType.IDENTIFIER -> parseTree.content
+            TokenType.FOREIGN_NAME -> {
+                val identifier = parseTree.content.substring(1, parseTree.content.lastIndex)
+                if (asIdentifier) {
+                    val walk = RegexDfa(TokenRegexParser.parseStringToRegex("""\l[\l\u\d_]*""")).newWalk()
```

Comment:
Would be more efficient to only call this `parseStringToRegex` once.

Reply:
Right.